quinn-proto/src/connection/assembler.rs

```diff
@@ -99,32 +107,60 @@ impl Assembler {
         let new = BinaryHeap::with_capacity(self.data.len());
         let old = mem::replace(&mut self.data, new);
         for chunk in old.into_sorted_vec().into_iter().rev() {
+            if chunk.size >= chunk.allocation_size {
```

Comment:
I also think this misses to split and append the last `buffer` segment?

Reply:
It will be `true` for a `Buffer` from a previous defragmentation, because every `Buffer` created inside the defragmentation method has `allocation_size == size`.

The split and append of the last `buffer` segment happens after the loop and is not affected by these changes.

I like your idea of skipping defragmentation for a buffer with a small proportional overhead, but I also want to hear the opinion of @djc and @Ralith on this.

Reply:
A smaller threshold like 1.2 seems fine to me.

Reply:
Done. I moved the logic to decide if a `Buffer` is considered to be fragmented into `Buffer::new` and adjust `allocation_size` accordingly, because a `Buffer` that will never be defragmented should not contribute to the over-allocation. Also I would not be allowed to do `self.allocated = self.bufferred` in `defragment`, if I simply skipped over a buffer with `allocation_size > size`.

Reply:
Thanks. That part looks good to me now